### PR TITLE
Assert equal for tables and collections

### LINKED ISSUE #1076: Add ``assert_equal`` to TableCollection and TreeSequence
The '.equals`` method is very useful, but it doesn't tell us where the problem is when tables are not equal. We should add methods ``assert_equal`` which would initially look like 
```python
def assert_equal(self, other, ignore_provenance=...) # same signature as equals
    assert self.sequence_length == other.sequence_length
    assert self.nodes == other.nodes
    # etc
```
This would at least tell us which table the mismatch occurs in. We can imagine providing a more detailed picture of how tables differ at some point, but this could be added later as a refinement. This form ``assert_equal`` should integrate nicely with pytest, and clean up our testing code some more.

Seems like a straightforward win?

### PR DESCRIPTION
**Pull request by benjeffery**
WIP - Tables are done, Table Collection has a few options that I'm working on.

Fixes #1076

### CODE REVIEW

**Comment by benjeffery** [comment]
@petrelharp as JK is away can I get a review here?

**Comment by benjeffery** [comment]
@mergiyio rebase

**Comment by benjeffery** [comment]
@Mergifyio rebase

**Comment by benjeffery** [comment]
Coverage isn't updating for some reason.

File: python/tests/test_tables.py [hunk 1]
```diff
@@ -1009,7 +1010,162 @@ def test_set_with_optional_properties(self, codec):
...
+    @pytest.fixture
+    def table1(self, test_rows):
+        table1 = self.table_class()
+        for row in test_rows[:5]:
+            table1.add_row(**row)
+        return table1
+
+    def test_equal(self, table1, test_rows):
+        table2 = self.table_class()
+        for row in test_rows[:5]:
```
Comment (petrelharp) [inline python/tests/test_tables.py:1034]: isn't this depending on the number of rows being 5? e.g., shouldn't you replace `5` with `table1.num_rows`?

File: python/tests/test_tables.py [hunk 2]
```diff
@@ -1009,7 +1010,162 @@ def test_set_with_optional_properties(self, codec):
...
+                match=re.escape(
+                    f"self.{column_name2}={test_rows[4][column_name2]} "
+                    f"other.{column_name2}={test_rows[5][column_name2]}"
+                ),
+            ):
+                table1.assert_equals(table2)
+
+    def test_num_rows(self, table1, test_rows):
+        table2 = self.table_class()
+        for row in test_rows[:4]:
```
Comment (petrelharp) [inline python/tests/test_tables.py:1117]: `4` -> `table1.num_rows - 1`?

File: python/tests/test_tables.py [hunk 3]
```diff
@@ -1009,7 +1010,162 @@ def test_set_with_optional_properties(self, codec):
...
+            table2.add_row(**row)
+        with pytest.raises(
+            AssertionError,
+            match=f"{type(table1).__name__} number of rows differ: self=5 other=4",
```
Comment (petrelharp) [inline python/tests/test_tables.py:1121]: oh I see, many things are depending on the number 5. perhaps it's OK.

**Review by petrelharp — APPROVED** [review]
Looks great! Thanks!!

File: python/tests/test_tables.py (hunk 1 above)
Comment (benjeffery) [reply python/tests/test_tables.py:1034]: The tables have a fix number of rows (10).

**Comment by benjeffery** [comment]
> Looks great! Thanks!!
Thanks for the review!

**Pull request merged**